File: hooks/block-protected-nix-config.py
#!/usr/bin/env python3
"""PreToolUse hook: block edits to protected lines in configuration.nix."""

import json
import re
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent))
from nix_protected import PROTECTED_PATTERNS

CONFIG_FILENAME = "configuration.nix"


def is_config_file(file_path: str) -> bool:
    return file_path.endswith(CONFIG_FILENAME)


def has_protected_pattern(text: str) -> list[str]:
    matched = []
    for pattern in PROTECTED_PATTERNS:
        if re.search(pattern, text):
            matched.append(pattern)
    return matched


def check_edit(tool_input: dict) -> str | None:
    file_path = tool_input.get("file_path", "")
    if not is_config_file(file_path):
        return None

    old_string = tool_input.get("old_string", "")
    new_string = tool_input.get("new_string", "")

    old_matches = has_protected_pattern(old_string)
    if not old_matches:
        return None

    for pattern in old_matches:
        old_lines = [ln for ln in old_string.splitlines() if re.search(pattern, ln)]
        new_lines = [ln for ln in new_string.splitlines() if re.search(pattern, ln)]
        if old_lines != new_lines:
            return f"保護対象の設定行を変更しようとしています: {pattern}"

    return None


def check_write(tool_input: dict) -> str | None:
    file_path = tool_input.get("file_path", "")
    if not is_config_file(file_path):
        return None

    path = Path(file_path)
    if not path.exists():
        return None

    current_content = path.read_text()
    new_content = tool_input.get("content", "")

    for pattern in PROTECTED_PATTERNS:
        current_lines = [ln.strip() for ln in current_content.splitlines() if re.search(pattern, ln)]
        new_lines = [ln.strip() for ln in new_content.splitlines() if re.search(pattern, ln)]
        if current_lines and current_lines != new_lines:
            return f"保護対象の設定行が変更または削除されます: {pattern}"

    return None


def main() -> None:
    data = json.load(sys.stdin)
    tool_name = data.get("tool_name", "")
    tool_input = data.get("tool_input", {})

    if tool_name == "Edit":
        reason = check_edit(tool_input)
    elif tool_name == "Write":
        reason = check_write(tool_input)
    else:
        return

    if reason:
        json.dump({"decision": "block", "reason": reason}, sys.stdout)


if __name__ == "__main__":
    main()

File: hooks-rs/src/hooks/block_protected_nix_config.rs
use regex::Regex;

use crate::input::HookInput;
use crate::nix_protected::PROTECTED_PATTERNS;
use crate::output::{block, pass};

pub fn run(input: &HookInput) {
    let file_path = input.tool_input.file_path_resolved();
    if !file_path.ends_with("configuration.nix") {
        return;
    }

    let compiled: Vec<Regex> = PROTECTED_PATTERNS
        .iter()
        .filter_map(|p| Regex::new(p).ok())
        .collect();

    match input.tool_name.as_str() {
        "Edit" => check_edit(input, &compiled),
        "Write" => check_write(input, &compiled),
        _ => pass(),
    }
}

fn check_edit(input: &HookInput, patterns: &[Regex]) {
    let old_string = &input.tool_input.old_string;
    let new_string = &input.tool_input.new_string;

    let old_matches: Vec<usize> = patterns
        .iter()
        .enumerate()
        .filter(|(_, re)| re.is_match(old_string))
        .map(|(i, _)| i)
        .collect();

    if old_matches.is_empty() {
        return;
    }

    for &idx in &old_matches {
        let old_lines: Vec<&str> = old_string
            .lines()
            .filter(|l| patterns[idx].is_match(l))
            .collect();
        let new_lines: Vec<&str> = new_string
            .lines()
            .filter(|l| patterns[idx].is_match(l))
            .collect();
        if old_lines != new_lines {
            block(&format!(
                "保護対象の設定行を変更しようとしています: {}",
                PROTECTED_PATTERNS[idx]
            ));
            return;
        }
    }
}

fn check_write(input: &HookInput, patterns: &[Regex]) {
    let path = std::path::Path::new(input.tool_input.file_path_resolved());
    if !path.exists() {
        return;
    }
    let current = std::fs::read_to_string(path).unwrap_or_default();
    let new_content = &input.tool_input.content;

    for (idx, re) in patterns.iter().enumerate() {
        let current_lines: Vec<&str> = current
            .lines()
            .filter(|l| re.is_match(l))
            .map(|l| l.trim())
            .collect();
        let new_lines: Vec<&str> = new_content
            .lines()
            .filter(|l| re.is_match(l))
            .map(|l| l.trim())
            .collect();
        if !current_lines.is_empty() && current_lines != new_lines {
            block(&format!(
                "保護対象の設定行が変更または削除されます: {}",
                PROTECTED_PATTERNS[idx]
            ));
            return;
        }
    }
}
